File: report-generator/src/test/java/com/teamscale/report/util/BashFileSkippingInputStreamTest.java
package com.teamscale.report.util;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.jar.JarEntry;
import java.util.jar.JarInputStream;

import static org.assertj.core.api.Assertions.assertThat;

class BashFileSkippingInputStreamTest {

	@Test
	void testBashFileJar() throws IOException {
		ArrayList<String> filesInJar = getEntriesFromJarFile("spring-boot-executable-example.jar");
		assertThat(filesInJar).hasSize(110);
	}

	@Test
	void testNormalJar() throws IOException {
		ArrayList<String> filesInJar = getEntriesFromJarFile("normal.jar");
		assertThat(filesInJar).hasSize(284);
	}

	private ArrayList<String> getEntriesFromJarFile(String resourceName) throws IOException {
		InputStream inputStream = getClass().getResourceAsStream(resourceName);
		BashFileSkippingInputStream bashFileSkippingInputStream = new BashFileSkippingInputStream(inputStream);
		JarInputStream jarInputStream = new JarInputStream(bashFileSkippingInputStream);
		JarEntry entry;
		ArrayList<String> filesInJar = new ArrayList<>();
		while ((entry = jarInputStream.getNextJarEntry()) != null) {
			filesInJar.add(entry.getName());
		}
		return filesInJar;
	}
}

File: report-generator/src/test/kotlin/com/teamscale/report/util/BashFileSkippingInputStreamTest.kt
package com.teamscale.report.util

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import java.io.IOException
import java.util.jar.JarEntry
import java.util.jar.JarInputStream

internal class BashFileSkippingInputStreamTest {
	@Test
	@Throws(IOException::class)
	fun testBashFileJar() {
		val filesInJar = getEntriesFromJarFile("spring-boot-executable-example.jar")
		Assertions.assertThat(filesInJar).hasSize(110)
	}

	@Test
	@Throws(IOException::class)
	fun testNormalJar() {
		val filesInJar = getEntriesFromJarFile("normal.jar")
		Assertions.assertThat(filesInJar).hasSize(284)
	}

	@Throws(IOException::class)
	private fun getEntriesFromJarFile(resourceName: String): List<String> {
		val inputStream = javaClass.getResourceAsStream(resourceName)
		val bashFileSkippingInputStream = BashFileSkippingInputStream(inputStream!!)
		val jarInputStream = JarInputStream(bashFileSkippingInputStream)
		return generateSequence { jarInputStream.nextJarEntry }
			.map { it.name }
			.toList()
	}
}
